Fix the ConcurrentModificationException on the _trackEventsOutsideAltSlaMap in the EventProducer

somandal opened the pull request:
On enabling the additional warning log to track TopicPartitions seeing events outside alternate SLA, we saw an instance of STALLED partitions. Further investigation showed that the logging function in the EventProducer was throwing a ConcurrentModificationException under certain circumstances. When this occurs, the caller's onCompletion send callback isn't called. This can cause an issue where we never acknowledge the successful send, and the BMM flushless producer never commits a newer offset.

This PR fixes this in two ways:
- Make the _trackEventsOutsideAltSlaMap a ConcurrentHashMap
- Catch all exceptions that can be thrown by hashmap operations and log them rather than throwing.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

vmaheshw commented on datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:
```diff
@@ -272,20 +273,28 @@ private void performSlaRelatedLogging(DatastreamRecordMetadata metadata, long ev
+          _trackEventsOutsideAltSlaMap.clear();
+          _lastEventsOutsideAltSlaLogTimeMs = System.currentTimeMillis();
+        }
+      } catch (ConcurrentModificationException | NullPointerException | IllegalArgumentException | ClassCastException
```
ConcurrentHashMap will not throw ConcurrentModificationException. You can get rid of it.

somandal replied:
done

vmaheshw reviewed (commented):
Do you think there is a value in moving sendCallback.onCompletion to finally block in `onSendCallback`, to avoid any future bug in `checkpoint()` or `reportMetrics` does not impact the checkpoint. The reason is this is a critical piece for FlushlessEventProducer otherwise we will be sending duplicate data, till the issue is identified and fixed.

somandal commented:
> Do you think there is a value in moving sendCallback.onCompletion to finally block in `onSendCallback`, to avoid any future bug in `checkpoint()` or `reportMetrics` does not impact the checkpoint. The reason is this is a critical piece for FlushlessEventProducer otherwise we will be sending duplicate data, till the issue is identified and fixed.

As discussed offline, I've modified the code to catch all exceptions and wrap them in the sendFailureException and call the onCompletion in the finally block.
